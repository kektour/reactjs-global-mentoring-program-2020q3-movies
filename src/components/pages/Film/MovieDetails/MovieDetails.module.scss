@import 'src/styles/shared.scss';

.root {
  height: 540px;
  background-image: url('/showcaseBackground.jpg');
  background-size: cover;
  position: relative;

  &::after {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1;
    background: rgba(0, 0, 0, 0.8);
    box-shadow: inset 120px 100px 250px #000000, inset -120px -100px 250px #000000;
  }
}

.container {
  @extend %container;
  z-index: 2;
  position: relative;
}

.header {
  padding-top: 20px;
  margin-bottom: 40px;
}

.goBackButton {
  background: transparent;
  border: none;

  &:hover {
    cursor: pointer;
    & > i {
      color: #fff;
    }
  }

  &:focus {
    outline: none;
  }

  & > i {
    font-size: 20px;
    color: #f65261;
  }
}

.body {
  display: flex;
}

.image {
  width: 280px;
  height: 400px;
  margin-right: 60px;
}

.info {
  display: flex;
  flex-direction: column;
  color: #bababa;
}

.titleNRating {
  padding-top: 40px;
  display: flex;
  align-items: center;
  margin-bottom: 20px;
}

.title {
  margin: 0;
  padding: 0;
  font-weight: normal;
  font-size: 50px;
  letter-spacing: 2px;
  margin-right: 20px;
}

.rating {
  width: 60px;
  height: 60px;
  line-height: 60px;
  text-align: center;
  font-size: 25px;
  border: 1px solid #fff;
  border-radius: 50%;
}

.note {
  font-size: 18px;
  margin-bottom: 30px;
}

.releaseDateNRuntime {
  display: flex;
  color: #f65261;
  font-size: 25px;
  margin-bottom: 30px;
}

.releaseDate {
  margin-right: 40px;
}

.overview {
  padding-right: 150px;
  line-height: 20px;
  font-size: 18px;
}
