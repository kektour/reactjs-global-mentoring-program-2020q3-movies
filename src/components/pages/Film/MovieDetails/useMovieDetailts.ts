import { useDispatch } from 'react-redux';
import { moviesResetFilterActionCreator } from '../../../../store/movies';

type UseMovieDetailts = {
  resetFilter: () => void;
};

export const useMovieDetailts = (): UseMovieDetailts => {
  const dispatch = useDispatch();
  const resetFilter = () => dispatch(moviesResetFilterActionCreator());
  return {
    resetFilter,
  };
};
