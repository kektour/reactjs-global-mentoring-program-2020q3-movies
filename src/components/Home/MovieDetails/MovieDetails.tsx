import axios from 'axios';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { getColorByRating } from '../../../helpers/common';
import { Movie } from '../../../models/movie';
import { Header } from '../../shared/Header';
import styles from './MovieDetails.module.scss';
import { useMovieDetailts } from './useMovieDetailts';

type Props = {
  movieId: string;
};

export const MovieDetails: React.FC<Props> = ({ movieId }) => {
  const history = useHistory();
  const movieDetailts = useMovieDetailts();
  const [movie, setMovie] = useState<Movie | undefined>();

  useEffect(() => {
    axios
      .get<Movie>(`http://localhost:4000/movies/${movieId}`)
      .then((res) => setMovie(res.data))
      .catch(() => history.push('/'));
  }, [movieId, history]);

  useEffect(() => movie && window.scrollTo(0, 0), [movie]);

  const movieYear = useMemo(() => movie?.release_date.substring(0, 4), [movie]);
  const ratingColor = useMemo(() => (movie ? getColorByRating(movie.vote_average) : undefined), [movie]);

  const goHandleGoBack = useCallback(() => {
    movieDetailts.resetFilter();
    history.push('/');
  }, [history, movieDetailts]);

  return (
    <div className={styles.root}>
      <div className={styles.container}>
        {!movie ? null : (
          <React.Fragment>
            <div className={styles.header}>
              <Header>
                <button className={styles.goBackButton} onClick={goHandleGoBack}>
                  <i className="fa fa-search"></i>
                </button>
              </Header>
            </div>
            <div className={styles.body}>
              <img className={styles.image} src={movie.poster_path} alt="Not Found" />
              <div className={styles.info}>
                <div className={styles.titleNRating}>
                  <h1 className={styles.title}>{movie.title}</h1>
                  <div className={styles.rating} style={{ color: ratingColor }}>
                    {movie.vote_average}
                  </div>
                </div>
                {!!movie.tagline && <div className={styles.note}>{movie.tagline}</div>}
                <div className={styles.releaseDateNRuntime}>
                  <div className={styles.releaseDate}>{movieYear}</div>
                  {movie.runtime} min
                </div>
                <div className={styles.overview}>{movie.overview}</div>
              </div>
            </div>
          </React.Fragment>
        )}
      </div>
    </div>
  );
};
