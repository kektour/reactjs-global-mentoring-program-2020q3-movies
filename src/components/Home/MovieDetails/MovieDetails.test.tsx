import React from 'react';
import { render, waitForDomChange } from '@testing-library/react';
import axios from 'axios';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MovieDetails } from './MovieDetails';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;
const mockStore = configureStore();
const movie = {
  id: 337167,
  title: 'Fifty Shades Freed',
  tagline: "Don't miss the climax",
  vote_average: 6.1,
  vote_count: 1195,
  release_date: '2018-02-07',
  poster_path: 'https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg',
  overview:
    'Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.',
  budget: 55000000,
  revenue: 136906000,
  genres: ['Drama', 'Romance'],
  runtime: 106,
};

describe('MovieDetails', () => {
  const spyScrollTo = jest.fn();

  beforeAll(() => {
    Object.defineProperty(global.window, 'scrollTo', { value: spyScrollTo });
  });

  it('Should render component correctly', async () => {
    const store = mockStore({});
    mockedAxios.get.mockImplementationOnce(() => Promise.resolve({ data: movie }));
    const component = render(
      <Provider store={store}>
        <MovieDetails movieId={movie.id + ''} />
      </Provider>
    );
    await waitForDomChange();
    expect(component.asFragment()).toMatchSnapshot();
  });
});
