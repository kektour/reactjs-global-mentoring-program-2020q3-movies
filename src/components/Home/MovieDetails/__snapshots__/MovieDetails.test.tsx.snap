// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="container"
    >
      <div
        class="header"
      >
        <div
          class="root"
        >
          <div
            class="container"
          >
            <div
              class="header"
            >
              <p
                class="root"
              >
                <span>
                  netflix
                </span>
                roulette
              </p>
              <button
                class="goBackButton"
              >
                <i
                  class="fa fa-search"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="body"
      >
        <img
          alt="Not Found"
          class="image"
          src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
        />
        <div
          class="info"
        >
          <div
            class="titleNRating"
          >
            <h1
              class="title"
            >
              Fifty Shades Freed
            </h1>
            <div
              class="rating"
              style="color: green;"
            >
              6.1
            </div>
          </div>
          <div
            class="note"
          >
            Don't miss the climax
          </div>
          <div
            class="releaseDateNRuntime"
          >
            <div
              class="releaseDate"
            >
              2018
            </div>
            106 min
          </div>
          <div
            class="overview"
          >
            Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
